### PR TITLE
Add UperNet

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR adds the classic [UperNet](https://arxiv.org/abs/1807.10221) framework to Transformers.

Many papers that introduce a new vision backbone, such as BEiT, ConvNeXt, Swin,... benchmark their model on downstream tasks such as semantic segmentation and object detection. All of these papers use the UperNet framework (introduced in 2018) when evaluating their backbone on semantic segmentation.

Hence, this PR implements this framework, making use of the new [AutoBackbone API](#20229) to make the following possible:

```
from transformers import SwinConfig, UperNetConfig, UperNetForSemanticSegmentation

backbone_config = SwinConfig(out_features=["stage1", "stage2", "stage3", "stage4"])

config = UperNetConfig(backbone_config=backbone_config)
model = UperNetForSemanticSegmentation(config)
```

In the code above, we're instantiating the UperNet framework with Swin Transformer as backbone. The code looks equivalent for another backbone, like ConvNeXt:
```
from transformers import ConvNextBackbone, UperNetConfig, UperNetForSemanticSegmentation

backbone_config = ConvNextBackbone(out_features=["stage1", "stage2", "stage3", "stage4"])

config = UperNetConfig(backbone_config=backbone_config)
model = UperNetForSemanticSegmentation(config)
```

To do:

- [ ] looking into supporting `from_pretrained` of backbones => will be done in a follow-up PR
- [x] make sure UperNetImageProcessor does exact same preprocessing
- [x] make UperNetImageProcessor also take `segmentation_maps` as optional input
- [x] add image processor tests
- [x] convert all checkpoints + update organization
- [x] fix integration tests

### CODE REVIEW

File: src/transformers/models/auto/configuration_auto.py [hunk 1]
```diff
@@ -474,6 +475,7 @@
         ("ul2", "UL2"),
         ("unispeech", "UniSpeech"),
         ("unispeech-sat", "UniSpeechSat"),
+        ("upernet", "UPerNet"),
```
Comment (sgugger) [inline src/transformers/models/auto/configuration_auto.py:478]: There should also be a line with the archive constant.

File: src/transformers/models/donut/modeling_donut_swin.py [hunk 2]
```diff
@@ -523,7 +523,7 @@ def __init__(self, config, dim, input_resolution, num_heads, shift_size=0):
         self.shift_size = shift_size
         self.window_size = config.window_size
         self.input_resolution = input_resolution
-        self.set_shift_and_window_size(input_resolution)
+        # self.set_shift_and_window_size(input_resolution)
```
Comment (sgugger) [inline src/transformers/models/donut/modeling_donut_swin.py:526]: Not sure this should be in this PR?

File: src/transformers/models/swin/modeling_swin.py [hunk 3]
```diff
@@ -589,7 +590,7 @@ def __init__(self, config, dim, input_resolution, num_heads, shift_size=0):
         self.shift_size = shift_size
         self.window_size = config.window_size
         self.input_resolution = input_resolution
-        self.set_shift_and_window_size(input_resolution)
+        # self.set_shift_and_window_size(input_resolution)
```
Comment (sgugger) [inline src/transformers/models/swin/modeling_swin.py:593]: Same comment here?

File: src/transformers/models/upernet/configuration_upernet.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+logger = logging.get_logger(__name__)
+
+
+class UperNetConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of an [`UperNetForSemanticSegmentation`]. It is used to
+    instantiate an UperNet model according to the specified arguments, defining the model architecture. Instantiating a
+    configuration with the defaults will yield a similar configuration to that of the UperNet
+    [bert-base-uncased](https://huggingface.co/bert-base-uncased) architecture.
```
Comment (sgugger) [inline src/transformers/models/upernet/configuration_upernet.py:33]: Checkpoint name is wrong here.

File: src/transformers/models/upernet/configuration_upernet.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+    Configuration objects inherit from [`PretrainedConfig`] and can be used to control the model outputs. Read the
+    documentation from [`PretrainedConfig`] for more information.
+
+    Args:
+        backbone_config (`PretrainedConfig`, *optional*, defaults to `ResNetConfig`):
```
Comment (sgugger) [inline src/transformers/models/upernet/configuration_upernet.py:39]:
```suggestion
        backbone_config (`PretrainedConfig`, *optional*, defaults to `ResNetConfig()`):
```

File: src/transformers/models/upernet/configuration_upernet.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        self.auxiliary_concat_input = auxiliary_concat_input
+        self.loss_ignore_index = loss_ignore_index
+
+    def to_dict(self):
+        """
+        Serializes this instance to a Python dictionary. Override the default [`~PretrainedConfig.to_dict`]. Returns:
+            `Dict[str, any]`: Dictionary of all the attributes that make up this configuration instance,
+        """
+        output = copy.deepcopy(self.__dict__)
+        print("Backbone config: ", type(self.backbone_config))
```
Comment (sgugger) [inline src/transformers/models/upernet/configuration_upernet.py:122]: To remove.

File: src/transformers/models/upernet/image_processing_upernet.py [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+            The scaling factor or maximum size.
+
+            If it is a float number, then the image will be rescaled by this factor, else if it is a tuple of 2
+            integers, then the image will be rescaled as large as possible within the scale.
+        return_scale (bool):
+            Whether to return the scaling factor besides the rescaled image size.
+    Returns:
+        tuple[int]: The new rescaled image size.
+    """
+    w, h = old_size
```
Comment (sgugger) [inline src/transformers/models/upernet/image_processing_upernet.py:73]: Use width, height if possible.

File: src/transformers/models/upernet/modeling_upernet.py [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+from .configuration_upernet import UperNetConfig
+
+
+UPERNET_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "facebook/upernet-convnext-tiny",
+    # See all UperNet models at https://huggingface.co/models?filter=upernet
+]
+
+
+class ConvModule(nn.Module):
```
Comment (sgugger) [inline src/transformers/models/upernet/modeling_upernet.py:36]: Needs to be prefixed with `UperNet`

File: src/transformers/models/upernet/modeling_upernet.py [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        output = self.conv(input)
+        output = self.batch_norm(output)
+        output = self.activation(output)
+
+        return output
+
+
+class PyramidPoolingBlock(nn.Module):
```
Comment (sgugger) [inline src/transformers/models/upernet/modeling_upernet.py:71]: Same here.

File: src/transformers/models/upernet/modeling_upernet.py [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+            self.add_module(str(i), layer)
+
+    def forward(self, input: torch.Tensor) -> torch.Tensor:
+        hidden_state = input
+        for layer in self.layers:
+            hidden_state = layer(hidden_state)
+        return hidden_state
+
+
+class PyramidPoolingModule(nn.Module):
```
Comment (sgugger) [inline src/transformers/models/upernet/modeling_upernet.py:88]: Same here.

File: src/transformers/models/upernet/modeling_upernet.py [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+
+    Args:
+        pool_scales (tuple[int]):
+            Pooling scales used in Pooling Pyramid Module.
+        in_channels (int):
+            Input channels.
+        channels (int):
+            Channels after modules, before conv_seg.
+        align_corners (bool):
+            align_corners argument of F.interpolate.
```
Comment (sgugger) [inline src/transformers/models/upernet/modeling_upernet.py:100]: Docstring should use our usual syntax.

File: src/transformers/models/upernet/modeling_upernet.py [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+        for ppm in self.blocks:
+            ppm_out = ppm(x)
+            upsampled_ppm_out = nn.functional.interpolate(
+                ppm_out, size=x.size()[2:], mode="bilinear", align_corners=self.align_corners
+            )
+            ppm_outs.append(upsampled_ppm_out)
+        return ppm_outs
+
+
+class UperHead(nn.Module):
```
Comment (sgugger) [inline src/transformers/models/upernet/modeling_upernet.py:126]: Need to be prefixed with `UperNet`.

File: src/transformers/models/upernet/modeling_upernet.py [hunk 13]
```diff
@@ -0,0 +1,365 @@
...
+
+    # TODO look into weights initialization
+    def _init_weights(self, module):
+        """Initialize the weights"""
+        if isinstance(module, BackboneMixin):
+            module.init_weights()
+        elif isinstance(module, UperHead):
+            module.init_weights()
+        elif isinstance(module, FCNHead):
+            module.init_weights()
```
Comment (sgugger) [inline src/transformers/models/upernet/modeling_upernet.py:293]: Those two are not `PreTrainedModel` so you can't call this method on them.

**Review by sgugger — COMMENTED** [review]
Thanks for working on this new model. Left a couple of comments.

File: src/transformers/models/upernet/modeling_upernet.py (hunk 13 above)
Comment (NielsRogge) [reply src/transformers/models/upernet/modeling_upernet.py:293]: Ok, however isolating `init_weights` to only be applied on the backbone module does not result in appropriate weights initialization.

When running `pytest tests/models/upernet/test_modeling_upernet.py::UperNetModelTest::test_initialization`, one gets:
```
AssertionError: -0.00017914299678523093 not found in [0.0, 1.0] : Parameter backbone.embeddings.patch_embeddings.weight of model <class 'transformers.models.upernet.modeling_upernet.UperNetForSemanticSegmentation'> seems not properly initialized
```

File: src/transformers/models/auto/configuration_auto.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/auto/configuration_auto.py:478]: Could you clarify what you mean by "archive constant"?

File: src/transformers/models/upernet/modeling_upernet.py (hunk 13 above)
Comment (NielsRogge) [reply src/transformers/models/upernet/modeling_upernet.py]: This is fixed now!

File: docs/source/en/model_doc/upernet.mdx [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+<small> UPerNet framework. Taken from the <a href="https://arxiv.org/abs/1807.10221">original paper</a>. </small>
+
+This model was contributed by [nielsr](https://huggingface.co/nielsr). The original code is based on OpenMMLab's mmsegmentation [here](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/decode_heads/uper_head.py).
+
+## Usage
+
+UPerNet is a general framework for semantic segmentation. It can be used with any vision backbone, like so:
+
+```
```
Comment (sgugger) [inline docs/source/en/model_doc/upernet.mdx:35]:
```suggestion
```py
```

File: docs/source/en/model_doc/upernet.mdx [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+backbone_config = SwinConfig(out_features=["stage1", "stage2", "stage3", "stage4"])
+
+config = UperNetConfig(backbone_config=backbone_config)
+model = UperNetForSemanticSegmentation(config)
+```
+
+To use another vision backbone, like [ConvNeXt](convnext), simply instantiate the model with the appropriate backbone:
+
+```
```
Comment (sgugger) [inline docs/source/en/model_doc/upernet.mdx:46]:
```suggestion
```py
```

File: src/transformers/models/upernet/configuration_upernet.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+        if backbone_config is None:
+            backbone_config = ResNetConfig()
+        else:
+            # verify that the backbone is supported
+            backbone_model_type = (
+                backbone_config.pop("model_type") if isinstance(backbone_config, dict) else backbone_config.model_type
+            )
+            if isinstance(backbone_config, dict):
+                config_class = CONFIG_MAPPING[backbone_model_type]
+                backbone_config = config_class.from_dict(backbone_config)
```
Comment (sgugger) [inline src/transformers/models/upernet/configuration_upernet.py:102]: Maybe raise an error if then the type is not a `PretrainedConfig`?

File: src/transformers/models/auto/configuration_auto.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/models/auto/configuration_auto.py:478]: [This one](https://github.com/huggingface/transformers/blob/1543cee7c8c95ef47f832b1f37625ba2923c4994/src/transformers/models/auto/configuration_auto.py#L186)

File: utils/check_repo.py [hunk 17]
```diff
@@ -674,9 +674,11 @@ def find_all_documented_objects():
     "TensorFlowBenchmark",
     "TensorFlowBenchmarkArguments",
     "BitBackbone",
+    "ConvNextBackbone",
     "MaskFormerSwinBackbone",
     "ResNetBackbone",
     "AutoBackbone",
+    "ConvNextBackbone",
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply utils/check_repo.py]: I'll keep the backbones undocumented for now but will remove the duplication here

File: src/transformers/models/upernet/modeling_upernet.py [hunk 18]
```diff
@@ -0,0 +1,380 @@
...
+    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained
+    models.
+    """
+
+    config_class = UperNetConfig
+    main_input_name = "pixel_values"
+    supports_gradient_checkpointing = True
+
+    # TODO look into weights initialization
+    def _init_weights(self, module):
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/upernet/modeling_upernet.py:301]: Could you clarify? Do you mean just `def init_weights` without `module` argument?

File: src/transformers/models/auto/configuration_auto.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/auto/configuration_auto.py:479]: Thanks, fixed

File: src/transformers/models/upernet/modeling_upernet.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/models/upernet/modeling_upernet.py:301]: I believe that is the definition of overwriting a method in Python (or most programmation languages), yes.

[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/upernet/modeling_upernet.py:301]: Yes but we still want the initialization of weights in `UperNetPreTrainedModel` to happen recursively, right?

If the `init_weights` method in this class won't take a `module` argument, then how can we check whether a submodule is an instance of the `BackboneMixin` class for instance?

File: src/transformers/models/upernet/modeling_upernet.py [hunk 19]
```diff
@@ -0,0 +1,380 @@
...
+    """
+    An abstract class to handle weights initialization and a simple interface for downloading and loading pretrained
+    models.
+    """
+
+    config_class = UperNetConfig
+    main_input_name = "pixel_values"
+    supports_gradient_checkpointing = True
+
+    # TODO look into weights initialization
```
Comment (alaradirik) [inline src/transformers/models/upernet/modeling_upernet.py:300]: Are the initialization settings of the modules finalized?

File: src/transformers/models/upernet/convert_convnext_upernet_to_pytorch.py [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+    url = "https://huggingface.co/datasets/hf-internal-testing/fixtures_ade20k/resolve/main/ADE_val_00000001.jpg"
+    image = Image.open(requests.get(url, stream=True).raw).convert("RGB")
+
+    # TODO: remove
+    # from torchvision.transforms import Compose, Normalize, Resize, ToTensor
+    # from transformers.utils.constants import IMAGENET_DEFAULT_MEAN, IMAGENET_DEFAULT_STD
+    # image_transforms = Compose(
+    #     [Resize((512, 512)), ToTensor(), Normalize(mean=IMAGENET_DEFAULT_MEAN, std=IMAGENET_DEFAULT_STD)]
+    # )
+    # pixel_values = image_transforms(image).unsqueeze(0)
```
Comment (alaradirik) [inline src/transformers/models/upernet/convert_convnext_upernet_to_pytorch.py:153]: Can be removed now:
```suggestion
```

File: src/transformers/models/upernet/convert_swin_upernet_to_pytorch.py [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+    url = "https://huggingface.co/datasets/hf-internal-testing/fixtures_ade20k/resolve/main/ADE_val_00000001.jpg"
+    image = Image.open(requests.get(url, stream=True).raw).convert("RGB")
+
+    # TODO: remove
+    # from torchvision.transforms import Compose, Normalize, Resize, ToTensor
+    # from transformers.utils.constants import IMAGENET_DEFAULT_MEAN, IMAGENET_DEFAULT_STD
+    # image_transforms = Compose(
+    #     [Resize((512, 512)), ToTensor(), Normalize(mean=IMAGENET_DEFAULT_MEAN, std=IMAGENET_DEFAULT_STD)]
+    # )
+    # pixel_values = image_transforms(image).unsqueeze(0)
```
Comment (alaradirik) [inline src/transformers/models/upernet/convert_swin_upernet_to_pytorch.py:231]: Same here
```suggestion
```

**Review by alaradirik — APPROVED** [review]
Thanks for adding this!

Left a few minor comments but everything looks good apart from the issues/comments related to configuration and model parameter initialisation (+ organization name update) and works as expected.

File: src/transformers/models/upernet/modeling_upernet.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/models/upernet/modeling_upernet.py:301]: By calling the right attributes of `self`. You can extract the base model using the prefix (if it exists) and then just initialize the proper subbmodules directly.

File: src/transformers/models/upernet/__init__.py [hunk 22]
```diff
@@ -0,0 +1,71 @@
+# flake8: noqa
+# There's no way to ignore "F401 '...' imported but unused" warnings in this
+# module, but to preserve other warnings. So, don't check this module at all.
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/upernet/__init__.py:3]: I think it's used in each init file, so I guess yes

File: src/transformers/models/upernet/modeling_upernet.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (NielsRogge) [reply src/transformers/models/upernet/modeling_upernet.py]: I made an update in https://github.com/huggingface/transformers/pull/20648/commits/3e2282bb809d2a3cbd3092eb500a109c4aeeacb0, let me know whether this is fine!

File: src/transformers/models/upernet/modeling_upernet.py (hunk 19 above)
Comment (NielsRogge) [reply src/transformers/models/upernet/modeling_upernet.py]: See comment above!

File: src/transformers/models/upernet/configuration_upernet.py [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+        auxiliary_num_convs=1,
+        auxiliary_concat_input=False,
+        loss_ignore_index=255,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+
+        if backbone_config is None:
+            logger.info("`backbone_config` is `None`. Initializing the config with the default `ResNet` backbone.")
+            backbone_config = ResNetConfig(out_features=["stage1", "stage2", "stage3", "stage4"])
```
Comment (sgugger) [inline src/transformers/models/upernet/configuration_upernet.py:95]: Small nit here, consider replacing `ResNetConfig` by `CONFIG_MAPPING["resnet"]` to remove the dependency on the resnet model.

File: src/transformers/models/upernet/image_processing_upernet.py [hunk 24]
```diff
@@ -0,0 +1,466 @@
...
+
+if is_torch_available():
+    import torch
+
+
+logger = logging.get_logger(__name__)
+
+
+# Copied from transformers.models.segformer.image_processing_segformer.SegformerImageProcessor with Segformer->UperNet
+class UperNetImageProcessor(BaseImageProcessor):
```
Comment (sgugger) [inline src/transformers/models/upernet/image_processing_upernet.py:51]: If it's a full copy, let's just reuse `SegformerImageProcessor` for this model by setting the right mapping in `image_processing_auto` (which needs to be done anyway).

**Review by sgugger — COMMENTED** [review]
LGTM on the model initialization. One last comment on the image processor: it needs to be added in the auto mapping, and since it seems to be a full copy of the SegformerImageProcessor, you should re-use it in the auto-mapping and not introduce a new image processor here.

File: src/transformers/models/upernet/configuration_upernet.py (hunk 23 above)
Comment (NielsRogge) [reply src/transformers/models/upernet/configuration_upernet.py:95]: Could you clarify why the dependency is to be avoided?

Wouldn't it be easier for readers to see we're instantiating a `ResNetConfig` by default (which is harder to figure out from the CONFIG_MAPPING)?

File: src/transformers/models/upernet/image_processing_upernet.py (hunk 24 above)
Comment (NielsRogge) [reply src/transformers/models/upernet/image_processing_upernet.py]: Ok, done!

File: src/transformers/models/upernet/configuration_upernet.py (hunk 23 above)
Comment (sgugger) [reply src/transformers/models/upernet/configuration_upernet.py:95]: Model files should not depend on each other, this is the core philosophy of Transformers. The line above explicitly says you are instantiating a resnet config so I don't think it makes the code harder to read.

**Review by sgugger — APPROVED** [review]
Thanks for all your work on this model!

**Comment by NielsRogge** [comment]
Thanks for the review, I'm waiting for the authors to respond regarding the creation of an organization on the hub.

File: src/transformers/models/upernet/configuration_upernet.py (hunk 23 above)
Comment (NielsRogge) [reply src/transformers/models/upernet/configuration_upernet.py]: Ok, I've updated it!

**Pull request merged**